src/backend/drm/renderer/mod.rs

```diff
@@ -0,0 +1,2460 @@
+                    }),
+                }),
+            event,
+        )
+    }
+}
+
+/// Possible buffers to export as a framebuffer using [`ExportFramebuffer`]
+#[derive(Debug)]
+pub enum ExportBuffer<'a, B: Buffer> {
+    /// A wayland buffer
+    Wayland(&'a WlBuffer),
+    /// A [`Allocator`] buffer
+    Allocator(&'a B),
+}
+
+impl<'a, B: Buffer> From<&'a UnderlyingStorage> for ExportBuffer<'a, B> {
+    fn from(storage: &'a UnderlyingStorage) -> Self {
+        match storage {
+            UnderlyingStorage::Wayland(buffer) => Self::Wayland(buffer),
+        }
+    }
+}
+
+/// Export a [`ExportBuffer`] as a framebuffer
+pub trait ExportFramebuffer<B: Buffer>
+where
+    B: Buffer,
+{
+    /// Type of the framebuffer
+    type Framebuffer: AsRef<framebuffer::Handle>;
+
+    /// Type of the error
+    type Error: std::error::Error;
+
+    /// Add a framebuffer for the specified buffer
+    fn add_framebuffer(
+        &self,
+        drm: &DrmDeviceFd,
+        buffer: ExportBuffer<'_, B>,
+    ) -> Result<Option<Self::Framebuffer>, Self::Error>;
+}
+
+impl<F, B> ExportFramebuffer<B> for Arc<Mutex<F>>
+where
+    F: ExportFramebuffer<B>,
+    B: Buffer,
+{
+    type Framebuffer = <F as ExportFramebuffer<B>>::Framebuffer;
+    type Error = <F as ExportFramebuffer<B>>::Error;
+
+    fn add_framebuffer(
+        &self,
+        drm: &DrmDeviceFd,
+        buffer: ExportBuffer<'_, B>,
+    ) -> Result<Option<Self::Framebuffer>, Self::Error> {
+        let guard = self.lock().unwrap();
+        guard.add_framebuffer(drm, buffer)
+    }
+}
+
+impl<F, B> ExportFramebuffer<B> for Rc<RefCell<F>>
+where
+    F: ExportFramebuffer<B>,
+    B: Buffer,
+{
+    type Framebuffer = <F as ExportFramebuffer<B>>::Framebuffer;
+    type Error = <F as ExportFramebuffer<B>>::Error;
+
+    fn add_framebuffer(
+        &self,
+        drm: &DrmDeviceFd,
+        buffer: ExportBuffer<'_, B>,
+    ) -> Result<Option<Self::Framebuffer>, Self::Error> {
+        self.borrow().add_framebuffer(drm, buffer)
+    }
+}
+
+type Frame<A, F> = FrameState<
+    DrmScanoutBuffer<
+        <A as Allocator>::Buffer,
+        <F as ExportFramebuffer<<A as Allocator>::Buffer>>::Framebuffer,
+    >,
+>;
+
+type FrameErrorType<A, F> = FrameError<
+    <A as Allocator>::Error,
+    <<A as Allocator>::Buffer as AsDmabuf>::Error,
+    <F as ExportFramebuffer<<A as Allocator>::Buffer>>::Error,
+>;
+
+type FrameResult<T, A, F> = Result<T, FrameErrorType<A, F>>;
+
+type RenderFrameErrorType<A, F, R> = RenderFrameError<
+    <A as Allocator>::Error,
+    <<A as Allocator>::Buffer as AsDmabuf>::Error,
+    <F as ExportFramebuffer<<A as Allocator>::Buffer>>::Error,
+    R,
+>;
+
+type RenderFrameResult<A, F, R> =
+    Result<(Option<Vec<Rectangle<i32, Physical>>>, RenderElementStates), RenderFrameErrorType<A, F, R>>;
+
+/// Specialized renderer utilizing drm planes with support for direct scan-out
+#[derive(Debug)]
+pub struct DrmRenderer<A, F, U>
```

Comment:
Might be a little late in the process, but I am not a huge fan of everything being a `Renderer` in smithay, with `DrmRenderer`s using `MultiRenderers`, that are wrapping `Gles2Renderer`, aided by the `DamageTrackedRenderer`.

Maybe we could name this thing `DrmCompositor` instead? (Or think about other ways to make the naming a little more self-explanatory?)

Reply:
Tbh I am also not a fan of the renderer naming, but I am also pretty bad at naming things 😅 

Renaming it shouldn't be a big undertaking, it should only affect two commits. DrmCompositor is fine for me (everything else that I came up with would end in Manager, which is just as bad as Renderer)

Reply:
renamed to `DrmCompositor`